framework/src/systems/SystemBase.C

```diff
@@ -1367,33 +1393,44 @@ SystemBase::solutionState(const unsigned int state) const
                _solution_states.size() - 1,
                " is available.");
 
+  if (state == 0)
+    mooseAssert(_solution_states[0] == &solutionInternal(), "Inconsistent current solution");
+  else
+    mooseAssert(_solution_states[state] == &getVector(oldSolutionStateVectorName(state)),
+                "Inconsistent solution state");
+
   return *_solution_states[state];
 }
 
 NumericVector<Number> &
 SystemBase::solutionState(const unsigned int state)
 {
-  // Create up to the state requested if unavailable
-  if (state >= _solution_states.size())
-  {
-    _solution_states.resize(state + 1);
+  if (!hasSolutionState(state))
+    needSolutionState(state);
+  return *_solution_states[state];
+}
 
-    // The first three states (now, old, older) will point to the solutions in the libMesh system,
-    // which is why we are using the "internal" calls to these vectors. _solution_states will then
-    // be the forward facing access to these vectors
+void
+SystemBase::needSolutionState(const unsigned int state)
+{
+  if (hasSolutionState(state))
+    return;
+
+  _solution_states.resize(state + 1);
+
+  // The 0-th (current) solution state is owned by libMesh
+  if (!_solution_states[0])
     _solution_states[0] = &solutionInternal();
-    if (state > 0)
-      _solution_states[1] = &solutionOldInternal();
-    if (state > 1)
-      _solution_states[2] = &solutionOlderInternal();
-
-    // Create anything that is past older (state of 3+)
-    for (unsigned int i = 3; i <= state; ++i)
-      if (!_solution_states[i])
-        _solution_states[i] = &addVector("solution_state_" + std::to_string(i), true, GHOSTED);
-  }
+  else
+    mooseAssert(_solution_states[0] == &solutionInternal(), "Inconsistent current solution");
```

Comment:
Maybe `mooseAssert` -> `mooseError`

Reply:
Same thing here - this is checking against an internal API that isn't touched by the user. I'd prefer an assertion if possible